components/properties/src/runtime.rs

```diff
@@ -0,0 +1,563 @@
+            "Math" => Self::Math,
+            "Noncharacter_Code_Point" | "NChar" => Self::NoncharacterCodePoint,
+            "Pattern_Syntax" | "Pat_Syn" => Self::PatternSyntax,
+            "Pattern_White_Space" | "Pat_WS" => Self::PatternWhiteSpace,
+            "Quotation_Mark" | "QMark" => Self::QuotationMark,
+            "Radical" => Self::Radical,
+            "Regional_Indicator" | "RI" => Self::RegionalIndicator,
+            "Sentence_Terminal" | "STerm" => Self::SentenceTerminal,
+            "Soft_Dotted" | "SD" => Self::SoftDotted,
+            "Terminal_Punctuation" | "Term" => Self::TerminalPunctuation,
+            "Unified_Ideograph" | "UIdeo" => Self::UnifiedIdeograph,
+            "Uppercase" | "Upper" => Self::Uppercase,
+            "Variation_Selector" | "VS" => Self::VariationSelector,
+            "White_Space" | "space" => Self::WhiteSpace,
+            "XID_Continue" | "XIDC" => Self::XidContinue,
+            "XID_Start" | "XIDS" => Self::XidStart,
+            _ => return None,
+        };
+
+        Some(prop)
+    }
+
+    /// Returns a type capable of looking up values for this property, as long as it is an
+    /// [ECMA262-subset binary property][ecma].
+    ///
+    /// This handles every property required by ECMA262, except for:
+    ///
+    /// - `Script` and `General_Category`: handle these directly with [`maps::load_general_category()`] and
+    ///    [`maps::load_script()`].
+    ///    using property values parsed via [`GeneralCategory::get_name_to_enum_mapper()`] and [`Script::get_name_to_enum_mapper()`]
+    ///    if necessary.
+    /// - `Script_Extensions`: handle this directly using APIs from [`crate::script`], like [`script::load_script_with_extensions_unstable()`]
+    /// - `General_Category` mask values: Handle this alongside `General_Category` using [`GeneralCategoryGroup`],
+    ///    using property values parsed via [`GeneralCategoryGroup::get_name_to_enum_mapper()`] if necessary
+    /// - `Assigned`, `All`, and `ASCII` pseudoproperties: Handle these using their equivalent sets,
+    ///    see docs on [`Self::parse_ecma262_name()`] for more details.
+    /// - `General_Category` property values can themselves be treated like properties using a shorthand in ECMA262,
+    ///    simply create the corresponding `GeneralCategory` set.
+    ///
+    /// ```rust
+    /// use icu::properties::runtime::UnicodeProperty;
+    ///
+    /// let property = UnicodeProperty::parse_ecma262_name("Emoji").expect("parsing string failed");
+    /// let data = property.load_ecma262_binary_property_unstable(&icu_testdata::unstable()).expect("loading data failed");
+    /// let emoji = data.as_borrowed();
+    ///
+    /// assert!(emoji.contains('🔥'));  // U+1F525 FIRE
+    /// assert!(!emoji.contains('V'));
+    /// ```
+    ///
+    /// [ecma]: https://tc39.es/ecma262/#table-binary-unicode-properties
+    pub fn load_ecma262_binary_property_unstable<P>(
```

Comment:
Suggestion/Discussion: maybe it would be better as just

```rust
let p = icu_properties::sets::load_for_ecma262_strict("Pattern_Syntax")?;
let p = icu_properties::sets::load_for_ecma262_loose("Pattern_Syntax")?;
```

with no intermediate type. The intermediate doesn't serve much purpose, and I don't see the use case of wanting to store it or load it directly. For example, there's never a reason to write

```rust
let p = icu_properties::runtime::BinaryProperty::PatternSyntax.load_ecma262_binary_property_unstable(...)
```

because you should use the function API instead.

There's the issue of whether you have a binary or an enumerated property. However, I think this isn't a huge issue right now, because

1. Ecma262 supports only a very small number of enumerated properties
2. When enumerated properties show up in the regex, I believe they have different regex syntax, so we already know from the lexer whether we are expecting a binary or an enumerated

Reply:
Hmm, that's a good idea. I still think we should have the UnicodeProperty API, but one thing we could do is make the current UnicodeProperty API `pub(crate)` and move this specific API to `sets` for now.

The reason I think this is still useful, though, is that it's very convenient to be able to match on UnicodeProperty values.

But how about for now I make it private and just expose `load_for_ecma262`, with some comments about eventually making it public

Reply:
Also doing it as a sets API makes my job as FFI person easier :smile:

Reply:
Clarification for the assertion above: it appears that ]the regex grammar in ECMA-262](https://tc39.es/ecma262/#_ref_11068) does indeed have different handling for `\p{X}` and `\p{Y=X}`, with the caveats that in `\p{X}`, it checks if `X` is a General_Category value before binary.

There's also the caveat that in the new `/v` mode, some of the binary properties could be properties of strings. I think this is fine and means that we can add a new `icu_properties::sets::load_for_ecma402_with_strings(...)` that returns a UnicodeSetData (CodePointInversionListAndStringList)

Reply:
How does this look?